File: xtquant/factors/technical.py
"""
小天量化交易 - 技术指标因子库
包含常用技术分析因子
"""

import numpy as np
from typing import Optional
from collections import deque

from .base import BaseFactor
from ..core.event import MarketEvent, EventType


class PriceFactor(BaseFactor):
    """价格因子 - 记录最新价格"""
    def __init__(self):
        super().__init__("price")
        self._prices: deque = deque(maxlen=1000)

    def calculate(self, event: MarketEvent) -> Optional[float]:
        if event.event_type in (EventType.TICK, EventType.BAR):
            price = event.data.get("price", event.data.get("close", 0))
            if price > 0:
                self._prices.append(price)
                return price
        return None


class VolumeFactor(BaseFactor):
    """成交量因子"""
    def __init__(self):
        super().__init__("volume")
        self._volumes: deque = deque(maxlen=1000)

    def calculate(self, event: MarketEvent) -> Optional[float]:
        if event.event_type in (EventType.TICK, EventType.BAR):
            vol = event.data.get("volume", event.data.get("vol", 0))
            if vol > 0:
                self._volumes.append(vol)
                return vol
        return None


class RSIFactor(BaseFactor):
    """RSI相对强弱指标"""
    def __init__(self, period: int = 14):
        super().__init__(f"RSI_{period}", {"period": period})
        self.period = period
        self._prices: deque = deque(maxlen=period + 1)

    def calculate(self, event: MarketEvent) -> Optional[float]:
        if event.event_type not in (EventType.TICK, EventType.BAR):
            return None
        price = event.data.get("price", event.data.get("close", 0))
        if price <= 0:
            return None

        self._prices.append(price)
        if len(self._prices) < self.period + 1:
            return None

        deltas = np.diff(list(self._prices))
        gains = deltas[deltas > 0]
        losses = -deltas[deltas < 0]
        avg_gain = np.mean(gains) if len(gains) > 0 else 0
        avg_loss = np.mean(losses) if len(losses) > 0 else 0

        if avg_loss == 0:
            return 100.0
        rs = avg_gain / avg_loss
        return 100.0 - (100.0 / (1.0 + rs))


class MACDFactor(BaseFactor):
    """MACD因子 - 返回MACD柱状图值"""
    def __init__(self, fast: int = 12, slow: int = 26, signal: int = 9):
        super().__init__(f"MACD_{fast}_{slow}", {"fast": fast, "slow": slow, "signal": signal})
        self.fast = fast
        self.slow = slow
        self.signal = signal
        self._prices: deque = deque(maxlen=slow + signal + 10)

    def _ema(self, data: np.ndarray, period: int) -> np.ndarray:
        alpha = 2.0 / (period + 1)
        ema = np.zeros_like(data)
        ema[0] = data[0]
        for i in range(1, len(data)):
            ema[i] = alpha * data[i] + (1 - alpha) * ema[i-1]
        return ema

    def calculate(self, event: MarketEvent) -> Optional[float]:
        if event.event_type not in (EventType.TICK, EventType.BAR):
            return None
        price = event.data.get("price", event.data.get("close", 0))
        if price <= 0:
            return None

        self._prices.append(price)
        if len(self._prices) < self.slow + self.signal:
            return None

        prices = np.array(list(self._prices))
        ema_fast = self._ema(prices, self.fast)
        ema_slow = self._ema(prices, self.slow)
        macd_line = ema_fast[-1] - ema_slow[-1]

        # 简化signal line计算
        macd_hist = prices[-self.signal:]
        signal_line = np.mean(macd_hist)  # 简化版，实际应做EMA

        return macd_line - signal_line


class OrderBookImbalanceFactor(BaseFactor):
    """订单簿失衡因子"""
    def __init__(self, depth: int = 5):
        super().__init__(f"OB_imbalance_{depth}", {"depth": depth})
        self.depth = depth

    def calculate(self, event: MarketEvent) -> Optional[float]:
        if event.event_type != EventType.ORDERBOOK:
            return None
        bids = event.data.get("bids", [])
        asks = event.data.get("asks", [])
        bid_vol = sum(float(q) for _, q in bids[:self.depth])
        ask_vol = sum(float(q) for _, q in asks[:self.depth])
        total = bid_vol + ask_vol
        return (bid_vol - ask_vol) / total if total > 0 else 0.0


class SpreadFactor(BaseFactor):
    """买卖价差因子 (bps)"""
    def __init__(self):
        super().__init__("spread_bps")

    def calculate(self, event: MarketEvent) -> Optional[float]:
        if event.event_type != EventType.ORDERBOOK:
            return None
        bids = event.data.get("bids", [])
        asks = event.data.get("asks", [])
        if not bids or not asks:
            return None
        bid = float(bids[0][0])
        ask = float(asks[0][0])
        mid = (bid + ask) / 2
        return (ask - bid) / mid * 10000 if mid > 0 else 0.0


class VWAPFactor(BaseFactor):
    """成交量加权平均价格因子"""
    def __init__(self, window: int = 20):
        super().__init__(f"VWAP_{window}", {"window": window})
        self.window = window
        self._prices: deque = deque(maxlen=window)
        self._volumes: deque = deque(maxlen=window)

    def calculate(self, event: MarketEvent) -> Optional[float]:
        if event.event_type not in (EventType.TICK, EventType.BAR):
            return None
        price = event.data.get("price", event.data.get("close", 0))
        volume = event.data.get("volume", event.data.get("vol", 0))
        if price <= 0 or volume <= 0:
            return None

        self._prices.append(price)
        self._volumes.append(volume)

        if len(self._prices) < self.window:
            return None

        prices = np.array(list(self._prices))
        volumes = np.array(list(self._volumes))
        return np.sum(prices * volumes) / np.sum(volumes)


class MomentumFactor(BaseFactor):
    """动量因子 - N期收益率"""
    def __init__(self, period: int = 20):
        super().__init__(f"momentum_{period}", {"period": period})
        self.period = period
        self._prices: deque = deque(maxlen=period + 1)

    def calculate(self, event: MarketEvent) -> Optional[float]:
        if event.event_type not in (EventType.TICK, EventType.BAR):
            return None
        price = event.data.get("price", event.data.get("close", 0))
        if price <= 0:
            return None

        self._prices.append(price)
        if len(self._prices) < self.period + 1:
            return None

        old_price = list(self._prices)[0]
        return (price - old_price) / old_price if old_price > 0 else 0.0


class VolatilityFactor(BaseFactor):
    """波动率因子 - N期收益率标准差"""
    def __init__(self, period: int = 20):
        super().__init__(f"volatility_{period}", {"period": period})
        self.period = period
        self._returns: deque = deque(maxlen=period)
        self._last_price: Optional[float] = None

    def calculate(self, event: MarketEvent) -> Optional[float]:
        if event.event_type not in (EventType.TICK, EventType.BAR):
            return None
        price = event.data.get("price", event.data.get("close", 0))
        if price <= 0:
            return None

        if self._last_price is not None and self._last_price > 0:
            ret = (price - self._last_price) / self._last_price
            self._returns.append(ret)

        self._last_price = price

        if len(self._returns) < self.period:
            return None

        return float(np.std(list(self._returns)))

File: gateway/internal/factor/technical.go
package factor

import (
	"math"

	"github.com/xiaotian-quant/gateway/internal/model"
)

// ── Price Factor ──

// PriceFactor tracks the latest close price.
type PriceFactor struct{ BaseFactor }

func NewPriceFactor() *PriceFactor {
	return &PriceFactor{NewBaseFactor("price")}
}

func (f *PriceFactor) Update(bar model.Bar) {
	f.BaseFactor.push(bar.Close)
}

// ── Volume Factor ──

// VolumeFactor tracks trading volume.
type VolumeFactor struct {
	BaseFactor
	prevVolume float64
}

func NewVolumeFactor() *VolumeFactor {
	return &VolumeFactor{BaseFactor: NewBaseFactor("volume")}
}

func (f *VolumeFactor) Update(bar model.Bar) {
	ratio := 0.0
	if f.prevVolume > 0 {
		ratio = bar.Volume / f.prevVolume
	}
	f.prevVolume = bar.Volume
	f.BaseFactor.push(ratio)
}

// ── RSI Factor (14-period) ──

// RSIFactor computes the Relative Strength Index.
type RSIFactor struct {
	BaseFactor
	period int
	gains  []float64
	losses []float64
	prevClose float64
}

func NewRSIFactor(period int) *RSIFactor {
	if period <= 0 {
		period = 14
	}
	return &RSIFactor{
		BaseFactor: NewBaseFactor("rsi"),
		period:     period,
	}
}

func (f *RSIFactor) Update(bar model.Bar) {
	if f.prevClose == 0 {
		f.prevClose = bar.Close
		return
	}
	change := bar.Close - f.prevClose
	f.prevClose = bar.Close

	gain, loss := 0.0, 0.0
	if change >= 0 {
		gain = change
	} else {
		loss = -change
	}

	f.gains = append(f.gains, gain)
	f.losses = append(f.losses, loss)
	if len(f.gains) > f.period*2 {
		f.gains = f.gains[len(f.gains)-f.period*2:]
		f.losses = f.losses[len(f.losses)-f.period*2:]
	}

	if len(f.gains) < f.period {
		return
	}

	avgGain := average(f.gains[len(f.gains)-f.period:])
	avgLoss := average(f.losses[len(f.losses)-f.period:])

	if avgLoss == 0 {
		f.BaseFactor.push(100)
		return
	}
	rs := avgGain / avgLoss
	rsi := 100 - (100 / (1 + rs))
	f.BaseFactor.push(rsi)
}

// ── MACD Factor (12, 26, 9) ──

// MACDFactor computes MACD line, signal line, and histogram.
type MACDFactor struct {
	BaseFactor
	fastPeriod   int
	slowPeriod   int
	signalPeriod int
	prices       []float64
	signalLine   float64
	histogram    float64
}

func NewMACDFactor(fast, slow, signal int) *MACDFactor {
	if fast <= 0 {
		fast = 12
	}
	if slow <= 0 {
		slow = 26
	}
	if signal <= 0 {
		signal = 9
	}
	return &MACDFactor{
		BaseFactor:   NewBaseFactor("macd"),
		fastPeriod:   fast,
		slowPeriod:   slow,
		signalPeriod: signal,
	}
}

func (f *MACDFactor) Update(bar model.Bar) {
	f.prices = append(f.prices, bar.Close)
	if len(f.prices) > f.slowPeriod+f.signalPeriod {
		f.prices = f.prices[len(f.prices)-f.slowPeriod-f.signalPeriod:]
	}

	if len(f.prices) < f.slowPeriod {
		return
	}

	fastEMA := ema(f.prices, f.fastPeriod)
	slowEMA := ema(f.prices, f.slowPeriod)
	macdLine := fastEMA - slowEMA

	signalLine := macdLine
	if f.signalLine != 0 {
		signalLine = f.signalLine + (macdLine-f.signalLine)*(2.0/(float64(f.signalPeriod)+1))
	}
	f.signalLine = signalLine
	f.histogram = macdLine - signalLine
	f.BaseFactor.push(macdLine)
}

// Histogram returns the MACD histogram.
func (f *MACDFactor) Histogram() float64 { return f.histogram }

// Signal returns the MACD signal line.
func (f *MACDFactor) Signal() float64 { return f.signalLine }

// ── OrderBook Imbalance Factor ──

// OBImbalanceFactor tracks the smoothed order book imbalance.
type OBImbalanceFactor struct {
	BaseFactor
	depth int
}

func NewOBImbalanceFactor(depth int) *OBImbalanceFactor {
	if depth <= 0 {
		depth = 5
	}
	return &OBImbalanceFactor{
		BaseFactor: NewBaseFactor("ob_imbalance"),
		depth:      depth,
	}
}

func (f *OBImbalanceFactor) UpdateFromOB(ob model.OrderBookData) {
	imb := ob.Imbalance(f.depth)
	f.BaseFactor.push(imb)
}

func (f *OBImbalanceFactor) Update(bar model.Bar) {
	// OB imbalance can only be updated from OrderBook events, not bars
}

// ── Spread Factor (bps) ──

// SpreadFactor tracks the bid-ask spread in basis points.
type SpreadFactor struct{ BaseFactor }

func NewSpreadFactor() *SpreadFactor {
	return &SpreadFactor{NewBaseFactor("spread_bps")}
}

func (f *SpreadFactor) UpdateFromOB(ob model.OrderBookData) {
	f.BaseFactor.push(ob.SpreadBps())
}

func (f *SpreadFactor) Update(bar model.Bar) {
	// Spread can only be updated from OrderBook events
}

// ── VWAP Factor (20-period) ──

// VWAPFactor computes the Volume-Weighted Average Price.
type VWAPFactor struct {
	BaseFactor
	period int
	prices []float64
	volumes []float64
}

func NewVWAPFactor(period int) *VWAPFactor {
	if period <= 0 {
		period = 20
	}
	return &VWAPFactor{
		BaseFactor: NewBaseFactor("vwap"),
		period:     period,
	}
}

func (f *VWAPFactor) Update(bar model.Bar) {
	f.prices = append(f.prices, bar.Close*bar.Volume)
	f.volumes = append(f.volumes, bar.Volume)

	if len(f.prices) > f.period {
		f.prices = f.prices[len(f.prices)-f.period:]
		f.volumes = f.volumes[len(f.volumes)-f.period:]
	}

	if len(f.prices) < f.period {
		return
	}

	priceVolSum := 0.0
	volSum := 0.0
	for i := 0; i < len(f.prices); i++ {
		priceVolSum += f.prices[i]
		volSum += f.volumes[i]
	}

	if volSum == 0 {
		return
	}
	vwap := priceVolSum / volSum

	// Store ratio of price to VWAP
	ratio := bar.Close / vwap
	f.BaseFactor.push(ratio)
}

// ── Momentum Factor (20-period) ──

// MomentumFactor computes price rate of change.
type MomentumFactor struct {
	BaseFactor
	period int
	prices []float64
}

func NewMomentumFactor(period int) *MomentumFactor {
	if period <= 0 {
		period = 20
	}
	return &MomentumFactor{
		BaseFactor: NewBaseFactor("momentum"),
		period:     period,
	}
}

func (f *MomentumFactor) Update(bar model.Bar) {
	f.prices = append(f.prices, bar.Close)
	if len(f.prices) > f.period+1 {
		f.prices = f.prices[len(f.prices)-f.period-1:]
	}

	if len(f.prices) < f.period+1 {
		return
	}

	prevPrice := f.prices[len(f.prices)-f.period-1]
	if prevPrice == 0 {
		return
	}
	momentum := (bar.Close - prevPrice) / prevPrice
	f.BaseFactor.push(momentum)
}

// ── Volatility Factor (20-period) ──

// VolatilityFactor computes annualized volatility.
type VolatilityFactor struct {
	BaseFactor
	period  int
	returns []float64
	prevPrice float64
}

func NewVolatilityFactor(period int) *VolatilityFactor {
	if period <= 0 {
		period = 20
	}
	return &VolatilityFactor{
		BaseFactor: NewBaseFactor("volatility"),
		period:     period,
	}
}

func (f *VolatilityFactor) Update(bar model.Bar) {
	if f.prevPrice > 0 {
		ret := (bar.Close - f.prevPrice) / f.prevPrice
		f.returns = append(f.returns, ret)
		if len(f.returns) > f.period {
			f.returns = f.returns[len(f.returns)-f.period:]
		}
	}
	f.prevPrice = bar.Close

	if len(f.returns) < 2 {
		return
	}

	avg := average(f.returns)
	variance := 0.0
	for _, r := range f.returns {
		variance += (r - avg) * (r - avg)
	}
	std := math.Sqrt(variance / float64(len(f.returns)-1))
	annualVol := std * math.Sqrt(252)
	f.BaseFactor.push(annualVol)
}

// ── Helpers ──

func average(vals []float64) float64 {
	if len(vals) == 0 {
		return 0
	}
	sum := 0.0
	for _, v := range vals {
		sum += v
	}
	return sum / float64(len(vals))
}

func ema(prices []float64, period int) float64 {
	if len(prices) < period {
		return 0
	}
	k := 2.0 / (float64(period) + 1)
	result := prices[len(prices)-period]
	for i := len(prices) - period + 1; i < len(prices); i++ {
		result = prices[i]*k + result*(1-k)
	}
	return result
}

// ── OBSpread Events ──

// UpdaterFromOB allows factors to be updated from order book data.
type UpdaterFromOB interface {
	UpdateFromOB(ob model.OrderBookData)
}

// UpdateAllFromOB calls UpdateFromOB on all factors that implement it.
func UpdateAllFromOB(pipeline *Pipeline, ob model.OrderBookData) {
	for _, f := range pipeline.factors {
		if updater, ok := f.(UpdaterFromOB); ok {
			updater.UpdateFromOB(ob)
		}
	}
}
